src/commonMain/kotlin/fr/acinq/lightning/crypto/LocalKeyManager.kt

```diff
@@ -3,168 +3,166 @@ package fr.acinq.lightning.crypto
-
-    override fun commitmentPoint(shaSeed: ByteVector32, index: Long): PublicKey = Generators.perCommitPoint(shaSeed, index)
-
-    override fun channelKeys(fundingKeyPath: KeyPath): ChannelKeys {
-        val fundingPubKey = fundingPublicKey(fundingKeyPath)
-        val recoveredChannelKeys = recoverChannelKeys(fundingPubKey.publicKey)
-        return ChannelKeys(
+    override fun channelKeys(fundingKeyPath: KeyPath): KeyManager.ChannelKeys {
+        // deterministic mode: use the funding pubkey to compute the channel key path
+        val fundingPrivateKey = privateKey(channelKeyBasePath I fundingKeyPath I hardened(0))
+        // we use the recovery process even in the normal case, which guarantees it works when we need it
+        val recoveredChannelKeys = recoverChannelKeys(fundingPrivateKey.publicKey())
+        return KeyManager.ChannelKeys(
             fundingKeyPath,
-            privateKey(fundingPubKey.path).privateKey,
-            recoveredChannelKeys.paymentKey,
-            recoveredChannelKeys.delayedPaymentKey,
-            recoveredChannelKeys.htlcKey,
-            recoveredChannelKeys.revocationKey,
-            recoveredChannelKeys.shaSeed
+            fundingPrivateKey = fundingPrivateKey,
+            paymentKey = recoveredChannelKeys.paymentKey,
+            delayedPaymentKey = recoveredChannelKeys.delayedPaymentKey,
+            htlcKey = recoveredChannelKeys.htlcKey,
+            revocationKey = recoveredChannelKeys.revocationKey,
+            shaSeed = recoveredChannelKeys.shaSeed
         )
     }
 
-    override fun recoverChannelKeys(fundingPubKey: PublicKey): RecoveredChannelKeys {
-        val channelKeyPath = KeyManager.channelKeyPath(fundingPubKey)
+    /**
+     * generate channel-specific keys and secrets (note that we cannot re-compute the channel's funding private key)
+     * @params fundingPubKey funding public key
+     * @return channel keys and secrets
+     */
+    fun recoverChannelKeys(fundingPubKey: PublicKey): RecoveredChannelKeys {
+        val channelKeyPrefix = channelKeyBasePath I channelKeyPath(fundingPubKey)
         return RecoveredChannelKeys(
             fundingPubKey,
-            paymentKey = privateKey(paymentPoint(channelKeyPath).path).privateKey,
-            delayedPaymentKey = privateKey(delayedPaymentPoint(channelKeyPath).path).privateKey,
-            htlcKey = privateKey(htlcPoint(channelKeyPath).path).privateKey,
-            revocationKey = privateKey(revocationPoint(channelKeyPath).path).privateKey,
-            shaSeed = shaSeed(channelKeyPath)
+            paymentKey = privateKey(channelKeyPrefix I hardened(2)),
+            delayedPaymentKey = privateKey(channelKeyPrefix I hardened(3)),
+            htlcKey = privateKey(channelKeyPrefix I hardened(4)),
+            revocationKey = privateKey(channelKeyPrefix I hardened(1)),
+            shaSeed = privateKey(channelKeyPrefix I hardened(5)).value.concat(1).sha256()
```

Comment:
Hardcoded-paths are all defined in this one place to minimize potential for footgun.